### PR TITLE
#2542 Dynamic proxy pattern

### PR DESCRIPTION
**Pull request by sashirestela**
Uploading the full structure for the Dynamic Proxy pattern:
- Java code
- Diagrams
- Readme

### CODE REVIEW

File: dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/AlbumInvocationHandler.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Class whose method 'invoke' will be called every time that an interface's method is called.
+ * That interface is linked to this class by the Proxy class.
+ */
+public class AlbumInvocationHandler implements InvocationHandler {
+
+  private static final Logger logger = LoggerFactory.getLogger(AlbumInvocationHandler.class);
```
Comment (ylcn91) [inline dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/AlbumInvocationHandler.java:40]: as the project already has Lombok support, so @Log can be used.

File: dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    }
+    var httpMethodName = httpMethodAnnotation.annotationType().getSimpleName().toUpperCase();
+    var url = baseUrl + buildUrl(method, args, httpMethodAnnotation);
+    var bodyPublisher = buildBodyPublisher(method, args);
+    var httpRequest = HttpRequest.newBuilder()
+        .uri(URI.create(url))
+        .header("Content-Type", "application/json")
+        .method(httpMethodName, bodyPublisher)
+        .build();
+    var httpResponse = httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofString());
```
Comment (ylcn91) [inline dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java:82]: HTTP error statuses (4xx, 5xx) are not explicitly handled here. Consider handling or logging them.

File: dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+  private String buildUrl(Method method, Object[] args, Annotation httpMethodAnnotation) {
+    var url = annotationValue(httpMethodAnnotation);
+    if (url == null) {
+      return "";
+    }
+    var index = 0;
+    for (var parameter : method.getParameters()) {
+      var pathAnnotation = getAnnotationOf(parameter.getDeclaredAnnotations(), Path.class);
+      if (pathAnnotation != null) {
+        var pathParam = "{" + annotationValue(pathAnnotation) + "}";
```
Comment (ylcn91) [inline dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java:95]: maybe we can consider encoding path parameters to handle special characters.

File: dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        return HttpRequest.BodyPublishers.ofString(body);
+      }
+      index++;
+    }
+    return HttpRequest.BodyPublishers.noBody();
+  }
+
+  private Object getResponse(Method method, HttpResponse<String> httpResponse) {
+    var rawData = httpResponse.body();
+    var returnType = method.getGenericReturnType();
```
Comment (ylcn91) [inline dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java:119]: Consider adding checks for safer casting.

File: dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+      Class<?> responseClass = (Class<?>) (((ParameterizedType) returnType)
+          .getActualTypeArguments()[0]);
+      return JsonUtil.jsonToList(rawData, responseClass);
+    } else {
+      Class<?> responseClass = method.getReturnType();
+      return JsonUtil.jsonToObject(rawData, responseClass);
+    }
+  }
+
+  private Annotation getHttpMethodAnnotation(Annotation[] annotations) {
```
Comment (ylcn91) [inline dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java:130]: This method could benefit from caching if the set of methods is limited to improve performance.

File: dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    var valueMethod = Arrays.stream(annotation.annotationType().getDeclaredMethods())
+        .filter(methodAnnot -> methodAnnot.getName().equals("value"))
+        .findFirst().orElse(null);
+    if (valueMethod == null) {
+      return null;
+    }
+    Object result;
+    try {
+      result = valueMethod.invoke(annotation, (Object[]) null);
+    } catch (Exception e) {
```
Comment (ylcn91) [inline dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java:152]: Consider logging this exception or rethrowing as a custom unchecked exception for better debugging.

File: dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/AlbumInvocationHandler.java (hunk 1 above)
Comment (sashirestela) [reply dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/AlbumInvocationHandler.java]: Done. `@Slf4j` was added

File: dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java (hunk 2 above)
Comment (sashirestela) [reply dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java:92]: Done.

File: dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java (hunk 3 above)
Comment (sashirestela) [reply dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java:111]: Done. Encoding for pathValue was added.

File: dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java (hunk 4 above)
Comment (sashirestela) [reply dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java]: Done. A check was added.

File: dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java (hunk 5 above)
Comment (sashirestela) [reply dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java]: Done. Local caching was added.

File: dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java (hunk 6 above)
Comment (sashirestela) [reply dynamic-proxy/src/main/java/com/iluwatar/dynamicproxy/tinyrestclient/TinyRestClient.java:175]: Done. Logging the exception.

**Comment by sashirestela** [comment]
@ylcn91 I've solved all of your observations.

**Comment by sashirestela** [comment]
@ylcn91 I was wondering if you need anything else from my side before you can review my fix commits. Please, let me know if I'm blocking you in any way.

**Review by iluwatar — DISMISSED** [review]

**Comment by iluwatar** [comment]
Looks good! Thank you for the contribution 🎉

@all-contributors please add @sashirestela for code

@all-contributors please add @ylcn91 for review

**Comment by iluwatar** [comment]
@all-contributors please add @ylcn91 for review

**Pull request merged**